Tools/RemoteSynchronization/Program.cs

```diff
@@ -0,0 +1,815 @@
+                    {
+                        Duplicati.Library.Logging.Log.WriteDryrunMessage(LOGTAG, "rsync", "Would delete {0} from {1}", f.Name, b.DisplayName);
+                    }
+                    else
+                    {
+                        await b.DeleteAsync(f.Name, CancellationToken.None);
+                    }
+                    successful_deletes++;
+                }
+                catch (Exception e)
+                {
+                    Duplicati.Library.Logging.Log.WriteErrorMessage(LOGTAG, "rsync", e, "Error deleting {0}: {1}", f.Name, e.Message);
+                }
+
+                i++;
+            }
+
+            sw.Stop();
+            Duplicati.Library.Logging.Log.WriteProfilingMessage(LOGTAG, "rsync", "Delete: {0} ms", sw.ElapsedMilliseconds);
+
+            if (GlobalConfig.Progress)
+                Console.WriteLine($"\rDeleting: {n}/{n}");
+            return successful_deletes;
+        }
+
+        /// <summary>
+        /// Creates an option that allows multiple tokens and multiple arguments per token.
+        /// </summary>
+        /// <param name="aliases">The aliases for the option.</param>
+        /// <param name="description">The description for the option.</param>
+        /// <returns>The created option.</returns>
+        private static Option<List<string>> OptionWithMultipleTokens(string[] aliases, string description)
+        {
+            return new Option<List<string>>(aliases: aliases, description: description)
+            {
+                Arity = ArgumentArity.OneOrMore,
+                AllowMultipleArgumentsPerToken = true
+            };
+        }
+
+        /// <summary>
+        /// Prepares the lists of files to copy, delete and verify.
+        /// The files to copy are the files that are not in the destination, have a different size or have a more recent modification date.
+        /// The files to delete are the files that are found in the destination but not found in the source.
+        /// The files to verify are the files that are found in both the source and the destination, and that have the same size and modification date.
+        /// </summary>
+        /// <param name="b_src">The source backend.</param>
+        /// <param name="b_dst">The destination backend.</param>
+        /// <param name="force">Flag for whether to force the synchronization.</param>
+        /// <returns>A tuple of Lists each holding the files to copy, delete and verify.</returns>
+        private static (IEnumerable<IFileEntry>, IEnumerable<IFileEntry>, IEnumerable<IFileEntry>) PrepareFileLists(IStreamingBackend b_src, IStreamingBackend b_dst)
+        {
+            var sw = new System.Diagnostics.Stopwatch();
+
+            sw.Restart();
+            var files_src = b_src.List();
+            sw.Stop();
+            Duplicati.Library.Logging.Log.WriteProfilingMessage(LOGTAG, "rsync", "Prepare | List source: {0} ms", sw.ElapsedMilliseconds);
+            sw.Restart();
```

Comment:
There is also a `Duplicati.Library.Logging.Timer` class you can use as:
```csharp
using (new Logging.Timer(LOGTAG, "rsync", "Prepare | List source"))
  var files_dst = b_dst.List();
```

It is not as precise as the stopwatch, but it has nanosecond resolution.

Reply:
Cool! I've implemented it where applicable. For some of the places, the timings are internal and accumulative, so there the Logging can't be used. But for all of the places where it was "start, run, stop, log" it makes a lot of sense! Only annoying thing is that the scope cannot be used to declare variables. Would be nice to have something like the `out` syntax, but that's for some other time!